backend/cn/simple_smt.ml

```diff
@@ -662,6 +662,44 @@ let get_expr s v: sexp =
   | Sexp.List [ Sexp.List [_;x] ] -> x
   | _ -> raise (UnexpectedSolverResponse res)
 
+
+let is_let (exp: sexp): (sexp StrMap.t * sexp) option =
+  match exp with
+  | Sexp.List [ Sexp.Atom "let"; Sexp.List binds; body ] ->
+    let add_bind mp bind =
+        match bind with
+        | Sexp.List [ Sexp.Atom x; y ] -> StrMap.add x y mp
+        | _ -> raise (UnexpectedSolverResponse bind)
+    in
+    let su = List.fold_left add_bind StrMap.empty binds in
+    Some (su, body)
+  | _ -> None
+
+(** Expand let-definitions in this term.
+ NOTE: this is intended to be used mostly on models generated by the
+ solver (e.g., `get-value` in Z3 sometimes contains `let`).  As such
+ we assume that `forall` and `exist` won't occur, and so we don't need
+ to check for variable capture. *)
```

Comment:
Won't the smt solver potentially return lambdas as models for arrays or functions, which we'd have to be careful about regarding variable capture?

Reply:
@yav

Reply:
It's hard to say as there is no standard for this, but in the cases I've seen arrays seem to be rendered as a bunch of concrete updates to a constant array.   It's probably worth trying to see if we can make it break though.  I am not even sure what notation the solvers might use for lambdas, the "standard" binders I see are `let`, `forall`, `exists`, and, I guess, the patterns in `match`, so I was thinking that I'd only expect to see `let` in a model.

Reply:
I think we should merge it as is.   Admittedly, it would be a nasty bug to track down, but it hard to fix without knowing what might be the binding construct...   I am pretty sure that `forall` and `exist` don't make in this context.

I see we have merge conflicts, so I'll resolve them, and then we can merge